### PR TITLE
feat: improve reference menus

### PR DESCRIPTION
**Pull request by Xazin**
Relates: #4300 #4179 #4031 #3841 

### Feature Preview

https://github.com/AppFlowy-IO/AppFlowy/assets/42929161/afeadfba-efa4-4590-9f5b-ce65f6a3e06a

---

#### PR Checklist

- [x] My code adheres to [AppFlowy's Conventions](https://docs.appflowy.io/docs/documentation/software-contributions/conventions)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [x] All existing tests are passing.

### CODE REVIEW

**Comment by Xazin** [comment]
The preview video is a bit outdated, it was before I enabled inserting pages from the inline page reference handler.

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/base/link_to_page_widget.dart [hunk 1]
```diff
@@ -1,252 +1,71 @@
...
 import 'package:easy_localization/easy_localization.dart';
-import 'package:flowy_infra_ui/style_widget/button.dart';
-import 'package:flowy_infra_ui/style_widget/text.dart';
-import 'package:flowy_infra_ui/widget/error_page.dart';
-import 'package:flutter/material.dart';
-import 'package:flutter/services.dart';
 
-void showLinkToPageMenu(
-  OverlayState container,
+InlineActionsMenuService? actionsMenuService;
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/base/link_to_page_widget.dart:12]:
```suggestion
InlineActionsMenuService? _actionsMenuService;
```

File: frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/base/link_to_page_widget.dart [hunk 2]
```diff
@@ -1,252 +1,71 @@
...
-    final acceptedKeys = [
-      LogicalKeyboardKey.arrowUp,
-      LogicalKeyboardKey.arrowDown,
-      LogicalKeyboardKey.tab,
-      LogicalKeyboardKey.enter,
-    ];
-
-    if (!acceptedKeys.contains(event.logicalKey)) {
-      return KeyEventResult.handled;
+  // ignore: use_build_context_synchronously
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/base/link_to_page_widget.dart:26]: No need to add the ignore rule.

File: frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart [hunk 3]
```diff
@@ -1,26 +1,51 @@
...
+  ///
+  final bool insertPage;
+
+  /// Defaults to 0 where there are no limits
+  /// Anything above 0 will limit the page reference results
+  /// to [limitResults].
+  ///
+  final int limitResults;
 
   late final ViewBackendService service;
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart:50]:
```
  final ViewBackendService service = ViewBackendService();
```

File: frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart [hunk 4]
```diff
@@ -29,9 +54,15 @@ class InlinePageReferenceService {
   Future<void> init() async {
     service = ViewBackendService();
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart:55]: Remove this line.

File: frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart [hunk 5]
```diff
@@ -29,9 +54,15 @@ class InlinePageReferenceService {
...
-      _filtered = value;
+
+      if (limitResults > 0) {
+        _filtered = value.take(limitResults).toList();
+      } else {
+        _filtered = value;
+      }
+
       _initCompleter.complete();
     });
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart:67]: An optional suggestion.

```
    _items = await _generatePageItems(currentViewId, viewLayout);
    _filtered = limitResults > 0 ? _items.take(limitResults).toList() : _items;
    _initCompleter.complete();
```

File: frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart [hunk 6]
```diff
@@ -40,19 +71,21 @@ class InlinePageReferenceService {
...
+          item.keywords != null &&
+          item.keywords!.isNotEmpty &&
+          item.keywords!.any(
+            (keyword) => keyword.contains(search.toLowerCase()),
+          ),
+    );
+
+    return limitResults > 0
+        ? items.take(limitResults).toList()
+        : items.toList();
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart:88]:
```
    final items = (search == null || search.isEmpty)
        ? _items
        : _items.where(
            (item) =>
                item.keywords != null &&
                item.keywords!.isNotEmpty &&
                item.keywords!.any(
                  (keyword) => keyword.contains(search.toLowerCase()),
                ),
          );

    return limitResults > 0
        ? items.take(limitResults).toList()
        : items.toList();
```

File: frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart [hunk 7]
```diff
@@ -61,15 +94,24 @@ class InlinePageReferenceService {
     _filtered = await _filterItems(search);
 
     return InlineActionsResult(
-      title: LocaleKeys.inlineActions_pageReference.tr(),
+      title: customTitle != null && customTitle!.isNotEmpty
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/plugins/inline_actions/handlers/inline_page_reference.dart:97]: `customTitle?.isNotEmpty == true`

**Pull request merged**